File: cs/2_2_decorator.cs
namespace Structural.Decorator
{
    interface IComponent
    {
        public abstract string Operation();
    }

    class ConcreteComponent : IComponent
    {
        public string Operation()
        {
            return "Concrete Component";
        }
    }

    class Decorator : IComponent
    {
        protected IComponent _component;
        public Decorator(IComponent component)
        {
            this._component = component;
        }
        public virtual string Operation()
        {
            return this._component.Operation();
        }

       
        public IComponent Undecorate(){
            return this._component;
        }
    }
    class ConcreteDecoratorA : Decorator
    {
        public ConcreteDecoratorA(IComponent comp) : base(comp)
        {
        }
        public override string Operation()
        {
            return $"ConcreteDecoratorA({base.Operation()})";
        }

        public int OtherOperation()
        {
            return base.Operation().Length;
        }
    }

    class ConcreteDecoratorB : Decorator
    {
        public ConcreteDecoratorB(IComponent comp) : base(comp)
        {
        }

        public override string Operation()
        {
            return $"ConcreteDecoratorB({base.Operation()})";
        }
    }
}

File: js/2_2_decorator.js
class ConcreteComponent {
  operation() {
    return "ConcreteComponent";
  }
}

class Decorator {
  constructor(component) {
    this._component = component;
  }

  operation() {
    if (this._component && this._component.operation) {
      return this._component.operation();
    } else {
      return "";
    }
  }

  undecorate() {
    return this._component;
  }
}

class ConcreteDecoratorA extends Decorator {
  operation() {
    return `ConcreteDecoratorA(${super.operation()})`;
  }

  otherOperation() {
    return super.operation().length;
  }
}

class ConcreteDecoratorB extends Decorator {
  operation() {
    return `ConcreteDecoratorB(${super.operation()})`;
  }
}


function DecoratorFunction(component) {
  function decorateMethod(method) {
    return function () {
      return `function decorator ${method.apply(component)}`;
    }
  }
  if (component && component.operation)
    component.operation = decorateMethod(component.operation);
  else
    component.operation = function () {
      return "";
    }
  component.otherOperation = function () {
    return this.operation().length;
  }
  return component;
}

export { ConcreteComponent, ConcreteDecoratorA, ConcreteDecoratorB, DecoratorFunction }